[Jia Wei] iP

LeeJiaWei23 opened the pull request:
### Clippy 📎 

> "The secret of getting ahead is getting started." - Mark Twain ([source](https://www.goodreads.com/quotes/219455-the-secret-of-getting-ahead-is-getting-started-the-secret))

Why trouble yourself to remember your daily tasks when you can get a someone else to do it.

- **command-line based**
- support multiple features
- EASY to pickup ~~and edit~~

To run this program,

1. download the files from [here](https://github.com/[LeeJiaWei23/ip](https://github.com/LeeJiaWei23/ip)).
2. run the program
3. add your tasks via commands
4. Take a break 

**And it is FREE!**

Features:

- [X] Add, Remove tasks
- [X] Local Copy of tasks
- [ ] Reminders (coming soon)
- [ ] Edit tasks (coming soon)
---
If you are familar with Java, you can use it to practice Java. Here's the `main` method

```java
public class Main {
    public static void main(String[] args) {
        App.start(Clippy.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

LeeJiaWei23 committed:
Add greeting

LeeJiaWei23 committed:
Add echo functionality

LeeJiaWei23 committed:
UI.java: Add helper class for text formatting

LeeJiaWei23 committed:
Add List.java

LeeJiaWei23 committed:
Add list functionality

LeeJiaWei23 committed:
Add functionality to mark and unmark tasks as done

LeeJiaWei23 committed:
Add support for ToDo tasks

LeeJiaWei23 committed:
Add suport for deadline tasks

LeeJiaWei23 committed:
Add support for event tasks

LeeJiaWei23 committed:
Clippy.java: Handle end-of-file exception in Clippy input

LeeJiaWei23 committed:
Add automated text UI testing

LeeJiaWei23 committed:
Clippy.java: Use ArrayList instead of fix array

LeeJiaWei23 committed:
Add exceptions to handle input errors

LeeJiaWei23 committed:
Add delete functionality and refactor code

LeeJiaWei23 committed:
Use Java enum for commands

LeeJiaWei23 committed:
Add file saving for task list

LeeJiaWei23 committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add file saving for task list

LeeJiaWei23 committed:
Support date time parsing

LeeJiaWei23 committed:
Fix storage bug for deleting task

LeeJiaWei23 committed:
Add find command to search tasks by date

LeeJiaWei23 committed:
Merge branch 'branch-Level-8'

LeeJiaWei23 committed:
Shift printing list logic to UI class

LeeJiaWei23 committed:
Organise classes into packages for better structure

LeeJiaWei23 committed:
Merge branch 'add-gradle-support'

LeeJiaWei23 committed:
Set up and run gradle

LeeJiaWei23 committed:
Format classes to pass style checks

LeeJiaWei23 committed:
Add JUnit Tests for Task and TaskList classes

LeeJiaWei23 committed:
Package the App as a JAR file

LeeJiaWei23 committed:
Add JavaDoc comments to classes methods

LeeJiaWei23 committed:
Update code to follow given coding standard

LeeJiaWei23 committed:
Rename Find command to Filter command

LeeJiaWei23 committed:
Add Find command to search keyword

LeeJiaWei23 committed:
Merge branch 'branch-A-JavaDoc'

LeeJiaWei23 committed:
Merge branch 'branch-A-CodingStandard'

LeeJiaWei23 committed:
Merge branch 'branch-Level-9'

gajinkee commented on src/main/java/clippy/parser/Parser.java:
```diff
@@ -0,0 +1,50 @@
+package clippy.parser;
+
+import clippy.ClippyException;
+import clippy.command.*;
```
Perhaps this should be expanded to individual imports instead of 1 wildcard?

gajinkee commented on src/main/java/clippy/parser/Parser.java:
```diff
@@ -0,0 +1,50 @@
+        String arguments = (words.length < 2 ? "" : words[1]);
+        CommandType commandType = getCommandType(words[0]);
+
+        return switch (commandType) {
+        case LIST -> new ListCommand();
+        case MARK -> new MarkCommand(arguments);
+        case UNMARK -> new UnmarkCommand(arguments);
+        case DELETE -> new DeleteCommand(arguments);
+        case TODO, DEADLINE, EVENT -> new AddCommand(commandType, input);
+        case BYE -> new ByeCommand();
+        case FILTER -> new FilterCommand(arguments);
+        case FIND -> new FindCommand(arguments);
+        };
```
Perhaps you could consider indenting for lambda style switch statements? Same for the other switch statements used.

gajinkee commented on src/test/java/clippy/task/TaskListTest.java:
```diff
@@ -0,0 +1,34 @@
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
+import static org.junit.jupiter.api.Assertions.assertThrows;
```
Should these static imports be placed at the top of imports list instead? I see the same in other files as well.

gajinkee commented on src/test/java/clippy/task/TaskTest.java:
```diff
@@ -0,0 +1,35 @@
+
+public class TaskTest {
+    private Task task;
+
+    @BeforeEach
+    void setUp() {
+        task = new ToDo("Sell Books");
+    }
+
+    @Test
+    void getStatusIcon_notDone_returnsEmptyBracket() {
+        assertEquals(" ", task.getStatusIcon());
+    }
+
+    @Test
+    void getStatusIcon_markAsDone_returnsX() {
+        task.markAsDone();
+        assertEquals("X", task.getStatusIcon());
+    }
+
+    @Test
+    void parseDate_validDateTime_success() throws ClippyException {
+        LocalDateTime expected = LocalDateTime.of(2025, 2, 1, 14, 30);
+        assertEquals(expected, task.parseDate("01/02/2025 1430"));
+    }
+}
```
Should these tests have JavaDocs as well?

gajinkee reviewed (commented):
LGTM. Just some nits to pick. 👍

LeeJiaWei23 committed:
Modify code to pass the checkstyle test

LeeJiaWei23 committed:
Add GUI to Clippy using JavaFX

LeeJiaWei23 committed:
Refine UI elements and style

LeeJiaWei23 committed:
Merge branch 'branch-Level-10'

pranavyey commented on src/main/java/clippy/Clippy.java:
```diff
@@ -0,0 +1,41 @@
+public class Clippy {
+    private final Storage storage;
+    private TaskList tasks;
+    private String commandType;
```
What is the purpose of commandType? Just curious

pranavyey commented on src/main/java/clippy/ClippyException.java:
```diff
@@ -0,0 +1,48 @@
+ * Represents exceptions specific to the Clippy application.
+ * This exception is thrown when an invalid operation or input is encountered.
+ */
+public class ClippyException extends Exception {
```
Good naming for the exceptions!

pranavyey commented on src/main/java/clippy/command/AddCommand.java:
```diff
@@ -0,0 +1,46 @@
+
+import clippy.ClippyException;
+import clippy.task.TaskList;
+import clippy.ui.UI;
```
Not sure about the naming as UI here

pranavyey commented on src/main/java/clippy/command/CommandType.java:
```diff
@@ -0,0 +1,18 @@
+ * Represents the different types of commands supported by the Clippy application.
+ * Each command type corresponds to a specific user action.
+ */
+public enum CommandType {
```
Perhaps can consider using switch and case due to the large number and potential expandability of command types?

pranavyey commented on src/main/java/clippy/command/FilterCommand.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    /**
+     * Executes the filter command by displaying tasks that match the given date.
```
Good explanation of the function!

pranavyey reviewed (commented):
Overall, good work! No coding standards violated(that I know of)!

LeeJiaWei23 committed:
Add assertions to improve code robustness

Application lacks assertions in key areas such as task management,
command parsing, and date handling.

Assertions help to catch logical errors early. This ensures that invalid
states are detected immediately during development.

Add assertion to:
* ensure task list is successful loaded from storage
* verify that parsed commands are never null before execution
* ensure task descriptions are non-empty when added
* check that task lists are properly initialized before listing

Using assertions are preferred over exception handling because they
are meant for detecting developer errors rather than user input errors.

Assertions are enabled with the '-ea' flag in Gradle

LeeJiaWei23 committed:
Implement auto-exit GUI when bye command is received

When the "bye" command is entered, Clippy does not
automatically close the application. User must manually
close which is not user-friendly.

Exiting automatically when "bye" is received improves the
overall user experience by ensuring a seamless shutdown of
the application.

Let's update the application to:
* modify `ByeCommand` to invoke `Platform.exit()` for GUI mode.
* ensure the JVM fully exits with `System.exit(0)`, preventing
  lingering background processes.

This ensures that both GUI and CLI modes exit cleanly when
the "bye" command is used, providing a more intuitive and
consistent user experience.

LeeJiaWei23 committed:
Refactor code in TaskList class

LeeJiaWei23 committed:
Merge pull request #1 from LeeJiaWei23/branch-A-Assertions

Add Assertions

LeeJiaWei23 committed:
Merge branch 'master' of https://github.com/LeeJiaWei23/ip

LeeJiaWei23 committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	data/tasks.txt

LeeJiaWei23 committed:
Merge pull request #2 from LeeJiaWei23/branch-A-CodeQuality

Branch code quality

LeeJiaWei23 committed:
Merge branch 'master' of https://github.com/LeeJiaWei23/ip

* 'master' of https://github.com/LeeJiaWei23/ip:
  Refactor code in TaskList class

LeeJiaWei23 committed:
Add undo command to remove previous command action

LeeJiaWei23 committed:
Add different font colors to improve GUI

LeeJiaWei23 committed:
Add more JUnit tests and JavaDoc comments

LeeJiaWei23 committed:
Update Clippy user icon

LeeJiaWei23 committed:
Add exception for argument-less commands

LeeJiaWei23 committed:
Update README.md

LeeJiaWei23 committed:
Update README.md

LeeJiaWei23 committed:
Add a screenshot of the product

LeeJiaWei23 committed:
Merge branch 'master' of https://github.com/LeeJiaWei23/ip

LeeJiaWei23 committed:
Update screenshot of product

LeeJiaWei23 committed:
Update README.md in docs directory

LeeJiaWei23 committed:
Edit code to pass checkstyle

LeeJiaWei23 committed:
Update README.md

LeeJiaWei23 committed:
Add exception message for redundant task status update

LeeJiaWei23 committed:
Fix empty indexStr bug in exception

LeeJiaWei23 committed:
Handle invalid data file format

LeeJiaWei23 committed:
Fix unclear exception messages

LeeJiaWei23 committed:
Fix event parsing order issue
